docs/core/diagnostics/diagnosticsource-diagnosticlistener.md

```diff
@@ -0,0 +1,413 @@
+
+MyListener TheListener = new MyListener();
+TheListener.Listening();
+
+HTTPClient Client = new HTTPClient();
+Client.SendWebRequest("https://docs.microsoft.com/dotnet/core/diagnostics/");
+
+class HTTPClient
+{
+    private static DiagnosticSource httpLogger = new DiagnosticListener("System.Net.Http");
+
+    public byte[] SendWebRequest(string url)
+    {
+        if (httpLogger.IsEnabled("RequestStart"))
+        {
+            httpLogger.Write("RequestStart", new { Url = url });
+        }
+        //Pretend this sends an HTTP request to the url and gets back a reply.
+        byte[] reply = new byte[] { };
+        return reply;
+
+    }
+}
+
+class Observer<T> : IObserver<T>
+{
+    public Observer(Action<T> onNext, Action onCompleted)
+    {
+        _onNext = onNext ?? new Action<T>(_ => { });
+        _onCompleted = onCompleted ?? new Action(() => { });
+    }
+
+    public void OnCompleted() { _onCompleted(); }
+    public void OnError(Exception error) { }
+    public void OnNext(T value) { _onNext(value); }
+
+    private Action<T> _onNext;
+    private Action _onCompleted;
+}
+
+class MyListener
+{
+    IDisposable networkSubscription;
+    IDisposable listenerSubscription;
+    private readonly object allListeners = new();
+    public void Listening()
+    {
+        Action<KeyValuePair<string, object>> whenHeard = delegate (KeyValuePair<string, object> data)
+        {
+            Console.WriteLine($"Data received: {data.Key}: {data.Value}");
+        };
+        Action<DiagnosticListener> onNewListener = delegate (DiagnosticListener listener)
+        {
+            Console.WriteLine($"New Listener discovered: {listener.Name}");
+            //Suscribe to the specific DiagnosticListener of interest.
+            if (listener.Name == "System.Net.Http")
+            {
+                //Use lock to ensure the callback code is thread safe.
+                lock(allListeners)
+                {
+                    if (networkSubscription != null)
+                    {
+                        networkSubscription.Dispose();
+                    }
+                    IObserver<KeyValuePair<string, object>> iobserver = new Observer<KeyValuePair<string, object>>(whenHeard, null);
+                    networkSubscription = listener.Subscribe(iobserver);
+                }
+                
+            }
+        };
+        //Subscribe to discover all DiagnosticListeners
+        IObserver<DiagnosticListener> observer = new Observer<DiagnosticListener>(onNewListener, null);
+        //When a listener is created, invoke the onNext function which calls the delegate.
+        listenerSubscription = DiagnosticListener.AllListeners.Subscribe(observer);
+
+    }
+    // Typically you leave the listenerSubscription subscription active forever.
+    // However when you no longer want your callback to be called, you can
+    // call listenerSubscription.Dispose() to cancel your subscription to the IObservable.
+}
+```
```

Comment:
I don't think it's necessary, but you could simplify this now that you've checked in the snippet as a separate program.

It would require two things:

1. Move the checked in .cs/.csproj pair to `docs/core/diagnostics/snippets/diagnosticsource/csharp`.
2. Add `// <WholeProgram>` to the beginning of the program and `// </WholeProgram>` at the end of the program.


Then you could replace this whole code snippet with:
```md
:::code language="csharp" source="snippets/diagnosticsource/csharp/Program.cs" id="WholeProgram":::
```

For any other snippets out of this code that you want to display, you just need to wrap the lines with
```csharp
// <Snippet1>
    var thisCodeIsInSnippet1 = 0;
// <Snippet 2>
    var thisCodeIsInSnippet1AndSnippet2 = 1;
// </Snippet2>
// </Snippet1>
```

Then, you could replace those code blocks with
```md
:::code language="csharp" source="snippets/diagnosticsource/csharp/Program.cs" id="Snippet1":::
:::code language="csharp" source="snippets/diagnosticsource/csharp/Program.cs" id="Snippet2":::
```

Again, I don't think it is absolutely necessary, but I thought I'd lay out how to do it.

Reply:
Agreed - we shouldn't duplicate the code inline and in a separate file.